R/osmdata.R

```diff
@@ -0,0 +1,241 @@
+  bb <- osmdata::getbb(city_name)
+  bb <- bb |> as.vector()
+  names(bb) <- c("xmin", "ymin", "xmax", "ymax")
+  bb <- sf::st_bbox(bb, crs = 4326)
+  return(bb)
+}
+
+#' Retrieve OpenStreetMap data for a given location
+#'
+#' Retrieve OpenStreetMap data for a given location, including
+#' the city boundary, the river centreline and surface, the streets, and the
+#' railways.
+#'
+#' @param city_name A character string with the name of the city.
+#' @param river_name A character string with the name of the river.
+#' @param crs An integer with the EPSG code for the projection. If no CRS is
+#'            specified, the default is the UTM zone for the city.
+#' @param buffer A numeric with the buffer distance in meters. By default,
+#'               no buffer is applied
+#'
+#' @return An list with the retrieved OpenStreetMap data sets for the
+#'         given location
+#' @export
+#'
+#' @examples
+#' get_osmdata("Bucharest", "Dambovita", buffer = 2000)
+get_osmdata <- function(city_name, river_name, crs = NULL, buffer = NULL) {
+  bb <- get_osm_bb(city_name)
+
+  if (is.null(crs)) crs <- get_utm_zone_epsg_bbox(bb) |> as.numeric()
+
+  if (!is.null(buffer)) {
+    bb <- bb |>
+      sf::st_as_sfc() |>
+      sf::st_transform(crs = crs) |>
+      sf::st_buffer(buffer) |>
+      sf::st_transform(crs = 4326) |>
+      sf::st_bbox()
+  }
+
+  boundary <- get_osm_city_boundary(city_name, bb, crs)
+  river <- get_osm_river(river_name, bb, crs)
+  srteets <- get_osm_streets(bb, crs)
+  railways <- get_osm_railways(bb, crs)
+
+  osm_data <- list(
+    bb = bb,
+    boundary = boundary,
+    river_centerline = river$centerline,
+    river_surface = river$surface,
+    streets = srteets,
+    railways = railways
+  )
+
+  return(osm_data)
+}
+
+#' Get the city boundary from OpenStreetMap
+#'
+#' This function retrieves the city boundary from OpenStreetMap based on a
+#' bounding box with the OSM tags "place:city" and "boundary:administrative".
+#' The result is filtered by the city name.
+#'
+#' @param city_name A character string with the name of the city
+#' @param bb Bounding box of class `bbox`
+#' @param crs Coordinate reference system as EPSG code
+#' @param multiple A logical indicating if multiple city boundaries should be
+#'                 returned. By default, only the first one is returned.
+#'
+#' @return An sf object with the city boundary
+#' @importFrom rlang .data
+#' @export
+#'
+#' @examples
+#' bb <- get_osm_bb("Bucharest")
+#' crs <- get_utm_zone_epsg_bbox(bb)
+#' get_osm_city_boundary("Bucharest", bb, crs)
+get_osm_city_boundary <- function(city_name, bb, crs, multiple = FALSE) {
+  # Define a helper function to fetch the city boundary
+  fetch_boundary <- function(key, value) {
+    CRiSp::osmdata_as_sf(key, value, bb)$osm_multipolygons |>
+      dplyr::filter(
+        .data$`name:en` == stringr::str_extract(city_name, "^[^,]+") |
+          .data$name == stringr::str_extract(city_name, "^[^,]+")
```

Comment:
Not sure I understand these comparisons: `city_name` is the input city name, was the regex match be intended for the column values?

Reply:
Ah, now I see it was to drop e.g. country name.